### PR TITLE
[FLINK-17599][docs] Update documents due to FLIP-84

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*This pr aims to update the documents due to [FLIP-84](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=134745878). Additional, add documents for USE/EXPLAIN/DESCRIBE statements, and do some code cleanup*

## Brief change log

  - *fix typos in TableEnvironment javadoc and PlannerBase javadoc*
  - *Code cleanup: use new methods introduced in FLIP-84 instead of deprecated methods*
  - *Update documents due to FLIP-84*
  - *Add documents for DESCRIBE statement*
  - *Add documents for EXPLAIN statement*
  - *Add documents for USE statement*

## Verifying this change

*(Please pick either of the following options)*

This change is document updating / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: docs/dev/table/sql/insert.zh.md [hunk 1]
```diff
@@ -40,12 +41,31 @@ EnvironmentSettings settings = EnvironmentSettings.newInstance()...
...
+tEnv.executeSql("CREATE TABLE GlassOrders(product VARCHAR, amount INT) WITH (...)");
+
+// 运行多个 INSERT 语句，将原表数据输出到多个结果表中
+StatementSet stmtSet = tEnv.createStatementSet();
+// `addInsertSql` 方法每次只接收单条 INSERT 语句
+stmtSet.addInsertSql(
+  "INSERT INTO RubberOrders SELECT product, amount FROM Orders WHERE product LIKE '%Rubber%'");
+stmtSet.addInsertSql(
+  "INSERT INTO GlassOrders SELECT product, amount FROM Orders WHERE product LIKE '%Glass%'");
+// 执行刚刚添加的所有的 INSERT 语句
```
Comment (leonardBang) [inline docs/dev/table/sql/insert.zh.md:64]:
```suggestion
// 执行刚刚添加的所有 INSERT 语句
```

File: docs/dev/table/sql/insert.zh.md [hunk 2]
```diff
@@ -40,12 +41,31 @@ EnvironmentSettings settings = EnvironmentSettings.newInstance()...
...
+TableResult tableResult2 = stmtSet.execute();
+// 可以通过 TableResult 来获取作业状态
```
Comment (leonardBang) [inline docs/dev/table/sql/insert.zh.md:66]:
```suggestion
// 通过 TableResult 来获取作业状态
```

File: docs/dev/table/sql/queries.md [hunk 3]
```diff
@@ -1,5 +1,5 @@
 ---
-title: "Queries"
+title: "SELECT queries"
```
Comment (leonardBang) [inline docs/dev/table/sql/queries.md:2]: In this section, Do we need change this? all queries should be SELECT queries from my understanding. Others should belong to other kind of SQL, "query" should not equals "SQL"

File: docs/dev/table/sql/queries.zh.md [hunk 4]
```diff
@@ -136,7 +136,78 @@ t_env.connect(FileSystem().path("/path/to/file")))
...
-    .sql_update("INSERT INTO RubberOrders SELECT product, amount FROM Orders WHERE product LIKE '%Rubber%'")
+    .execute_sql("INSERT INTO RubberOrders SELECT product, amount FROM Orders WHERE product LIKE '%Rubber%'")
+{% endhighlight %}
+</div>
+</div>
+
+{% top %}
+
+## Execute SELECT query
+SELECT 语句可以通过 `TableEnvironment.executeSql()` 方法来执行，将选择的结果收集到本地。该方法返回 `TableResult` 对象用于包装查询的结果。和 SELECT 语句很像，`Table.execute()` 方法可以将 `Table` 的内容收集到本地。
```
Comment (leonardBang) [inline docs/dev/table/sql/queries.zh.md:147]: 和 SELECT 语句类似

File: docs/dev/table/sql/queries.zh.md [hunk 5]
```diff
@@ -136,7 +136,78 @@ t_env.connect(FileSystem().path("/path/to/file")))
...
+`TableResult.collect()` 方法返回一个可以关闭的行迭代器。除非所有的数据都被搜集到本地，否则一个查询作业一直不会结束。所以我们应该通过 `CloseableIterator#close()` 方法主动的关闭作业以防止资源泄露。
```
Comment (leonardBang) [inline docs/dev/table/sql/queries.zh.md:148]: 搜集->收集， 
否则一个查询作业一直不会结束-> 否则一个查询作业永远不会结束
主动的 -> 主动地

File: docs/dev/table/sql/queries.zh.md [hunk 6]
```diff
@@ -136,7 +136,78 @@ t_env.connect(FileSystem().path("/path/to/file")))
...
+我们可以通过 `TableResult.print()` 方法将查询结果打印到本地控制台，但是我们应该确保最终返回的数据量比较小。因为所有数据先都收集到本地内存再执行打印操作。
```
Comment (leonardBang) [inline docs/dev/table/sql/queries.zh.md:149]: 因为所有数据都会先收集到本地内存再执行打印操作。

File: docs/dev/table/tableApi.md [hunk 7]
```diff
@@ -2120,13 +2120,13 @@ orders.insertInto("OutOrders")
...
-        <p>Similar to the INSERT INTO clause in a SQL query. Performs a insertion into a registered output table.</p>
+        <p>Similar to the INSERT INTO clause in a SQL query. Performs a insertion into a registered output table. The executeInsert method will immediately submit a flink job which execute the insert operation.</p>
 
         <p>Output tables must be registered in the TableEnvironment (see <a href="common.html#register-a-tablesink">Register a TableSink</a>). Moreover, the schema of the registered table must match the schema of the query.</p>
 
 {% highlight python %}
-orders = table_env.from_path("Orders");
-orders.insert_into("OutOrders");
+orders = table_env.from_path("Orders")
+orders.execute_insert("OutOrders")
```
Comment (leonardBang) [inline docs/dev/table/tableApi.md:2129]:
```suggestion
orders = table_env.from_path("Orders");
orders.execute_insert("OutOrders");
```

File: docs/dev/table/tableApi.zh.md [hunk 8]
```diff
@@ -2119,13 +2119,13 @@ orders.insertInto("OutOrders")
         <span class="label label-primary">批处理</span> <span class="label label-primary">流处理</span>
       </td>
       <td>
-        <p>类似于SQL请求中的INSERT INTO子句。将数据输出到一个已注册的输出表中。</p>
+        <p>类似于SQL请求中的INSERT INTO子句。将数据输出到一个已注册的输出表中。execute_insert 方法会立即提交一个 Flink 作业，触发插入操作。</p>
```
Comment (leonardBang) [inline docs/dev/table/tableApi.zh.md:2122]:
```suggestion
        <p>类似于SQL请求中的INSERT INTO子句。将数据输出到一个已注册的输出表中。`execute_insert` 方法会立即提交一个 Flink 作业，触发插入操作。</p>
```

File: docs/dev/table/common.md [hunk 9]
```diff
@@ -839,8 +828,14 @@ Table API and SQL queries are translated into [DataStream]({{ site.baseurl }}/de
...
+* `Table.execute()` is called. This method is used for collecting the table content to local client, and the Table API is translated immediately once this method is called.
+* `SatementSet.execute()` is called. A `Table` (emitted to a sink through `SatementSet.addInsert()`) or an INSERT statement (specified through `SatementSet.addInsertSql()`) will be buffered in `SatementSet` first. They are translated once `SatementSet.execute()` is called. All sinks will be optimized into one DAG.
 * A `Table` is translated when it is converted into a `DataStream` (see [Integration with DataStream and DataSet API](#integration-with-datastream-and-dataset-api)). Once translated, it's a regular DataStream program and is executed when `StreamExecutionEnvironment.execute()` is called.
+
+**Note:** Since 1.11 version, `sqlUpdate()` and `insertInto()` are deprecated. If the table program is built from these two methods, we must use `StreamTableEnvironment.execute()` method instead of `StreamExecutionEnvironment.execute()` method to execute it.
```
Comment (leonardBang) [inline docs/dev/table/common.md:837]: method `sqlUpdate()` and `insertInto()`

File: flink-end-to-end-tests/flink-batch-sql-test/src/main/java/org/apache/flink/sql/tests/BatchSQLTestProgram.java [hunk 10]
```diff
@@ -48,7 +48,7 @@
  *
  * <p>Parameters:
  * -outputPath output file path for CsvTableSink;
- * -sqlStatement SQL statement that will be executed as sqlUpdate
+ * -sqlStatement SQL statement that will be executed as executeSql
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-end-to-end-tests/flink-batch-sql-test/src/main/java/org/apache/flink/sql/tests/BatchSQLTestProgram.java:51]: Yes, I had replace all except `sqlUpdate` method in sql-client and `sqlUpdate` method in old planner.

Regarding (1), we can't replace `sqlUpdate` with `executeSql ` now, because we need to do more refactor work about this. (such as: submit job logic)

Regarding (2), them are in old planner

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/batch/table/TableSinkITCase.scala [hunk 11]
```diff
@@ -48,12 +48,10 @@ class TableSinkITCase extends BatchTestBase {
 
     registerCollection("MyTable", data3, type3, "a, b, c", nullablesOfData3)
 
-    tEnv.from("MyTable")
+    val table = tEnv.from("MyTable")
       .where('a > 20)
       .select("12345", 55.cast(DataTypes.DECIMAL(10, 0)), "12345".cast(DataTypes.CHAR(5)))
-      .insertInto("sink")
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/runtime/batch/table/TableSinkITCase.scala:54]: yes, we need replace all `Table.insertInto` methods with `Table.executeInsert`, `execInsertSqlAndWaitResult` is only used for testing.

File: docs/dev/table/sql/create.zh.md [hunk 12]
```diff
@@ -69,38 +69,38 @@ val tableEnv = TableEnvironment.create(settings)
...
 </div>
 
 <div data-lang="python" markdown="1">
 {% highlight python %}
-settings = EnvironmentSettings.newInstance()...
-table_env = TableEnvironment.create(settings)
+settings = EnvironmentSettings.new_instance()...
+table_env = StreamTableEnvironment.create(env, settings)
 
 # 对已经已经注册的表进行 SQL 查询
```
Comment (leonardBang) [inline docs/dev/table/sql/create.zh.md:91]: 对已注册的表进行 SQL 查询

File: docs/dev/table/sql/explain.zh.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+
+* This will be replaced by the TOC
+{:toc}
+
+EXPLAIN 语句用来解释一个 SELECT 语句或者 DML 的逻辑计划和优化后的计划。
+
+
+## 运行一个 EXPLAIN 语句
+
+EXPLAIN 语句可以通过 `TableEnvironment` 的 `executeSql()` 执行，也可以在 [SQL CLI]({{ site.baseurl }}/zh/dev/table/sqlClient.html) 中执行 EXPLAIN 语句。 若 EXPLAIN 操作执行成功，executeSql() 方法返回解释的结果，否则会抛出异常。
```
Comment (leonardBang) [inline docs/dev/table/sql/explain.zh.md:33]:
```suggestion
EXPLAIN 语句可以通过 `TableEnvironment` 的 `executeSql()` 执行，也可以在 [SQL CLI]({{ site.baseurl }}/zh/dev/table/sqlClient.html) 中执行 EXPLAIN 语句。 若 EXPLAIN 操作执行成功，`executeSql()` 方法返回解释的结果，否则会抛出异常。
```

File: docs/dev/table/sql/insert.zh.md [hunk 14]
```diff
@@ -40,12 +41,31 @@ EnvironmentSettings settings = EnvironmentSettings.newInstance()...
 TableEnvironment tEnv = TableEnvironment.create(settings);
 
 // 注册一个 "Orders" 源表，和 "RubberOrders" 结果表
-tEnv.sqlUpdate("CREATE TABLE Orders (`user` BIGINT, product VARCHAR, amount INT) WITH (...)");
-tEnv.sqlUpdate("CREATE TABLE RubberOrders(product VARCHAR, amount INT) WITH (...)");
+tEnv.executeSql("CREATE TABLE Orders (`user` BIGINT, product VARCHAR, amount INT) WITH (...)");
+tEnv.executeSql("CREATE TABLE RubberOrders(product VARCHAR, amount INT) WITH (...)");
 
 // 运行一个 INSERT 语句，将源表的数据输出到结果表中
```
Comment (leonardBang) [inline docs/dev/table/sql/insert.zh.md:47]: 运行一条

File: docs/dev/table/sql/insert.zh.md [hunk 15]
```diff
@@ -55,27 +75,66 @@ val settings = EnvironmentSettings.newInstance()...
 val tEnv = TableEnvironment.create(settings)
 
 // 注册一个 "Orders" 源表，和 "RubberOrders" 结果表
-tEnv.sqlUpdate("CREATE TABLE Orders (`user` BIGINT, product STRING, amount INT) WITH (...)")
-tEnv.sqlUpdate("CREATE TABLE RubberOrders(product STRING, amount INT) WITH (...)")
+tEnv.executeSql("CREATE TABLE Orders (`user` BIGINT, product STRING, amount INT) WITH (...)")
+tEnv.executeSql("CREATE TABLE RubberOrders(product STRING, amount INT) WITH (...)")
 
 // 运行一个 INSERT 语句，将源表的数据输出到结果表中
```
Comment (leonardBang) [inline docs/dev/table/sql/insert.zh.md:81]: 运行一条？or unify to `一个` I saw you used `执行一条` above

File: docs/dev/table/catalogs.zh.md [hunk 16]
```diff
@@ -100,10 +115,13 @@ Flink SQL> CREATE TABLE mytable (name STRING, age INT) WITH (...);
...
+</div>
+</div>
+
 
 更多详细信息，请参考[Flink SQL CREATE DDL]({{ site.baseurl }}/zh/dev/table/sql/create.html)。
 
-### 使用 Java/Scala/Python API
+### 使用 Java/Scala
 
 用户可以用编程的方式使用Java、Scala 或者 Python API 来创建 Catalog 表。
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply docs/dev/table/catalogs.zh.md:126]: Currently, we can't create a python catalog unless we use jvm_gateway, but that's a little hack. I get the confirm from @dianfu that we can delete `Python API` part now, @dianfu will re-add once his refactor work is done.

File: docs/dev/table/sql/insert.zh.md (hunk 15 above)
Comment (godfreyhe) [reply docs/dev/table/sql/insert.zh.md:81]: typo, we should use `执行一条`

File: docs/dev/table/sql/queries.md (hunk 3 above)
Comment (godfreyhe) [reply docs/dev/table/sql/queries.md]: change to `SELECT statements` which style is the same as others

File: docs/dev/table/sql/insert.md [hunk 17]
```diff
@@ -55,27 +75,65 @@ val settings = EnvironmentSettings.newInstance()...
...
+tEnv.executeSql("CREATE TABLE GlassOrders(product VARCHAR, amount INT) WITH (...)")
+
+// run multiple INSERT queries on the registered source table and emit the result to registered sink tables
+val stmtSet = tEnv.createStatementSet()
+// only single INSERT query can be accepted by `addInsertSql` method
+stmtSet.addInsertSql(
+  "INSERT INTO RubberOrders SELECT product, amount FROM Orders WHERE product LIKE '%Rubber%'")
+stmtSet.addInsertSql(
+  "INSERT INTO GlassOrders SELECT product, amount FROM Orders WHERE product LIKE '%Glass%'")
+// execute all statements as a batch
```
Comment (wuchong) [inline docs/dev/table/sql/insert.md:98]: `// execute all statements together`? `batch` may confuse users.

File: docs/dev/table/sql/queries.md [hunk 18]
```diff
@@ -134,16 +131,88 @@ t_env.connect(FileSystem().path("/path/to/file")))
...
 
 {% top %}
 
-## Supported Syntax
+## Execute SELECT statement
+A SELECT statement can be executed to collect the content to local through the `TableEnvironment.executeSql()` method. The method returns the result of the SELECT statement as a `TableResult`. Similar to SELECT statement, a `Table` can also be executed to collect to content of the table to local through the `Table.execute()` method.
+`TableResult.collect()` method returns a closeable row iterator. The select job will not be finished unless all result data has been collected. We should actively close the job to avoid resource leak through the `CloseableIterator#close()` method. 
+We can print the select result to client console through the `TableResult.print()` method. But we should make sure all result data to print should be small, because all data will be collected to local memory first, and then print them to console.
+
+**Notes:** Only append-only streaming SELECT statement is supported now.
```
Comment (wuchong) [inline docs/dev/table/sql/queries.md:148]:
```suggestion
**Notes:** For streaming mode, only append-only query is supported now.
```

File: docs/dev/table/sql/explain.md [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+[AS XML | AS JSON] 
+FOR 
+<select_statement_or_insert_statement>
+
+{% endhighlight %}
+
+For select syntax, please refer to [SELECT]({{ site.baseurl }}/dev/table/sql/queries.html#supported-syntax) page.
+For insert syntax, please refer to [INSERT]({{ site.baseurl }}/dev/table/sql/insert.html) page.
+
+Currently, only `EXPLAIN PLAN FOR <select_or_insert_statement>` is supported.
```
Comment (wuchong) [inline docs/dev/table/sql/explain.md:190]: I would prefer not list the unsupported syntax (`EXCLUDING ATTRIBUTES`, etc...), because it doesn't help anything for users, but confuse users. Then we can remove the above line.

File: docs/dev/table/sql/use.md [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+
+{% top %}
+
+## Syntax
+
+{% highlight sql %}
+USE CATALOG catalog_name
+
+USE [catalog_name.]database_name
+{% endhighlight %}
```
Comment (wuchong) [inline docs/dev/table/sql/use.md:191]: Split them into two section? `USE CATALOG`, `USE DATABASE`?  
And add a description under the syntax, e.g .

```
Set the current database. All subsequent commands that do not explicitly specify a database will use this one. If the provided database does not exist, an exception is thrown. The default current database is `default_database`.
```

File: docs/dev/table/sql/use.md [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+* This will be replaced by the TOC
+{:toc}
+
+USE statements are used to set the default database or catalog.
```
Comment (wuchong) [inline docs/dev/table/sql/use.md:28]: `default` -> `current`?

File: docs/dev/table/common.md [hunk 22]
```diff
@@ -1039,7 +1032,9 @@ val retractStream: DataStream[(Boolean, Row)] = tableEnv.toRetractStream[Row](ta
 </div>
 </div>
 
-**Note:** A detailed discussion about dynamic tables and their properties is given in the [Dynamic Tables](streaming/dynamic_tables.html) document.
+**Note:** A detailed discussion about dynamic tables and their properties is given in the [Dynamic Tables](streaming/dynamic_tables.html) document. 
+
+<span class="label label-danger">Attention</span> **Once the Table is converted to a DataStream, we must use the StreamExecutionEnvironment.execute method to execute the DataStream program.**
```
Comment (wuchong) [inline docs/dev/table/common.md:1037]:
```suggestion
<span class="label label-danger">Attention</span> **Once the Table is converted to a DataStream, please use the `StreamExecutionEnvironment.execute` method to execute the DataStream program.**
```

File: docs/try-flink/table_api.md [hunk 23]
```diff
@@ -123,11 +123,14 @@ tEnv.registerTableSource("transactions", new BoundedTransactionTableSource());
...
-tEnv
-    .scan("transactions")
-    .insertInto("spend_report");
-
-env.execute("Spend Report");
+// build table program
+Table table = tEnv.scan("transactions");
+
+// trigger execution
+TableResult tableResult = table.executeInsert("spend_report");
```
Comment (wuchong) [inline docs/try-flink/table_api.md:130]:
```
		TableResult tableResult = tEnv
				.scan("transactions")
				.executeInsert("spend_report");
```

Keep the same with archetype code? I also prefer this one because it's more concise for example.

File: docs/dev/table/sql/queries.md [hunk 24]
```diff
@@ -1,5 +1,5 @@
 ---
-title: "Queries"
+title: "SELECT statements"
```
Comment (wuchong) [inline docs/dev/table/sql/queries.md:2]: Remove `Operations` section as we moved them into separate pages.

File: docs/try-flink/table_api.md (hunk 23 above)
Comment (godfreyhe) [reply docs/try-flink/table_api.md:130]: There are `The Query` step and `Execute` step in `Breaking Down The Code` section. This design is for these steps.

File: docs/dev/table/sql/queries.md (hunk 24 above)
Comment (godfreyhe) [reply docs/dev/table/sql/queries.md:2]: I will revert `SELECT statements` to `Queries` because Queries contain not only SELECT statement but also VALUES statement

File: docs/try-flink/table_api.md [hunk 25]
```diff
@@ -462,14 +469,17 @@ public class SpendReport {
...
             .groupBy("accountId, w")
-            .select("accountId, w.start as timestamp, amount.sum")
-            .insertInto("spend_report");
+            .select("accountId, w.start as timestamp, amount.sum");
 
-        env.execute("Spend Report");
+        // trigger execution
+        TableResult tableResult = table.executeInsert("spend_report");
+        // wait job finished
+        tableResult.getJobClient().get()
```
Comment (twalthr) [inline docs/try-flink/table_api.md:481]: side comment: This API looks very ugly and we should not document it like this. I also saw in our tests and didn't like it. How about we introduce a `TableResult.awaitCompletion(timeout)`? For DCL and DDL, it would continue immediately.

File: docs/try-flink/table_api.zh.md [hunk 26]
```diff
@@ -451,6 +457,7 @@ import org.apache.flink.streaming.api.TimeCharacteristic;
 import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
 import org.apache.flink.table.api.Tumble;
 import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.TableResult;
```
Comment (twalthr) [inline docs/try-flink/table_api.zh.md:460]: use a `org.apache.flink.table.api.*` for all table examples to simplify the imports

File: docs/try-flink/table_api.md (hunk 25 above)
Comment (wuchong) [reply docs/try-flink/table_api.md:481]: Big +1 for this. We also have the same feeling and had some discussion around this. Maybe we can also have a `TableResult.awaitCompletion()` for this.

**Comment by wuchong** [comment]
+1 from my side when Timo's comments are addressed.

File: docs/try-flink/table_api.md (hunk 25 above)
Comment (godfreyhe) [reply docs/try-flink/table_api.md:481]: Thanks for the feedback. agree with you that current api is not friendly for testing or demos.  
For INSERT job, `TableResult.awaitCompletion(timeout)` will work fine. but for `SELECT` job, there are some problems: in current implementation, client will send a request to JM (which will re-forward the request to TM) to fetch a part of result when `Iterator.next()` is called. The sink operator has a fixed-length pool to buffer the result. If the pool is full, the back-pressure mechanism will trigger. So the job will not finish until all data has been fetched to client (or the job is canceled).  Code snippet:

```
TableResult tableResult = ...
TableResult.awaitCompletion(timeout); 
CloseableIterator<Row> it = TableResult.collect();
it...
```

`TableResult.awaitCompletion(timeout)` will block the execution of `TableResult.collect()`. This will cause the job will not finish, or get TimeoutException thrown by `awaitCompletion` method.

File: docs/dev/table/catalogs.md [hunk 27]
```diff
@@ -104,29 +119,67 @@ Flink SQL> CREATE TABLE mytable (name STRING, age INT) WITH (...);
...
+Users can use Java or Scala to create catalog tables programmatically.
 
 <div class="codetabs" markdown="1">
 <div data-lang="java" markdown="1">
 {% highlight java %}
 TableEnvironment tableEnv = ...
 
 // Create a HiveCatalog 
-Catalog catalog = new HiveCatalog("myhive", null, "<path_of_hive_conf>", "<hive_version>");
+Catalog catalog = new HiveCatalog("myhive", null, "<path_of_hive_conf>", "<hive_version>")
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply docs/dev/table/catalogs.md:138]: not intended, just typo

File: docs/try-flink/table_api.md (hunk 25 above)
Comment (twalthr) [reply docs/try-flink/table_api.md:481]: Yes, I think the behavior is fine. Maybe we can express it in a better name and definition. How about `TableResult.await()` that "waits until a result is ready"? Either the result is ready through `collect` or for INSERT INTO the result is ready once the job is completed and the number of rows is returned (at least this is the case for batch and would be the same for streaming in the future).

Comment (godfreyhe) [reply docs/try-flink/table_api.md:481]: make sense to me

**Comment by godfreyhe** [comment]
cc @twalthr

File: docs/dev/table/common.md [hunk 28]
```diff
@@ -849,18 +844,16 @@ Table API and SQL queries are translated into [DataStream]({{ site.baseurl }}/de
...
+A Table API or SQL query is translated when:
 
-For batch, a Table API or SQL query is translated when:
+* `TableEnvironment.executeSql()` is called. This method is used for executing a given statement, and the sql query is translated immediately once this method is called.
+* `Table.executeInsert()` is called. This method is used for inserting the table content to the given sink path, and the Table API is translated immediately once this method is called.
+* `Table.execute()` is called. This method is used for collecting the table content to local client, and the Table API is translated immediately once this method is called.
+* `SatementSet.execute()` is called. A `Table` (emitted to a sink through `SatementSet.addInsert()`) or an INSERT statement (specified through `SatementSet.addInsertSql()`) will be buffered in `SatementSet` first. They are translated once `SatementSet.execute()` is called. Each sink will be optimized independently. The execution graph contains multiple independent sub-DAGs.
```
Comment (twalthr) [inline docs/dev/table/common.md:852]: typo for `StatementSet` at multiple locations

File: docs/dev/table/sql/create.md [hunk 29]
```diff
@@ -48,16 +48,16 @@ TableEnvironment tableEnv = TableEnvironment.create(settings);
...
 // SQL query with a registered table
 // register a table named "Orders"
-tableEnv.sqlUpdate("CREATE TABLE Orders (`user` BIGINT, product STRING, amount INT) WITH (...)");
+tableEnv.executeSql("CREATE TABLE Orders (`user` BIGINT, product STRING, amount INT) WITH (...)");
 // run a SQL query on the Table and retrieve the result as a new Table
 Table result = tableEnv.sqlQuery(
   "SELECT product, amount FROM Orders WHERE product LIKE '%Rubber%'");
 
-// SQL update with a registered table
+// Execute insert sql with a registered table
```
Comment (twalthr) [inline docs/dev/table/sql/create.md:56]: use `SQL` instead of `sql` consistently in docs

File: docs/dev/table/sql/index.md [hunk 30]
```diff
@@ -28,7 +28,7 @@ This page describes the SQL language supported in Flink, including Data Definiti
 
 This page lists all the supported statements supported in Flink SQL for now:
 
-- [SELECT (Queries)](queries.html)
+- [Queries](queries.html)
```
Comment (twalthr) [inline docs/dev/table/sql/index.md:31]: can we revert this change? Having the keyword directly can help (`SELECT, VALUES`).

File: docs/dev/table/sql/queries.md [hunk 31]
```diff
@@ -134,16 +131,100 @@ t_env.connect(FileSystem().path("/path/to/file")))
...
+    .execute_sql("INSERT INTO RubberOrders SELECT product, amount FROM Orders WHERE product LIKE '%Rubber%'")
 {% endhighlight %}
 </div>
 </div>
 
 {% top %}
 
-## Supported Syntax
+## Execute a Query
+A SELECT statement or a VALUES statement can be executed to collect the content to local through the `TableEnvironment.executeSql()` method. The method returns the result of the SELECT statement (or the VALUES statement) as a `TableResult`. Similar to SELECT statement, a `Table` can also be executed to collect to content of the table to local through the `Table.execute()` method.
```
Comment (twalthr) [inline docs/dev/table/sql/queries.md:144]:
```suggestion
A SELECT statement or a VALUES statement can be executed to collect the content to local through the `TableEnvironment.executeSql()` method. The method returns the result of the SELECT statement (or the VALUES statement) as a `TableResult`. Similar to a SELECT statement, a `Table` object can be executed using the `Table.execute()` method to collect the content of the query to the local client.
```

File: docs/dev/table/sql/queries.md [hunk 32]
```diff
@@ -134,16 +131,100 @@ t_env.connect(FileSystem().path("/path/to/file")))
...
+`TableResult.collect()` method returns a closeable row iterator. The select job will not be finished unless all result data has been collected. We should actively close the job to avoid resource leak through the `CloseableIterator#close()` method. 
+We can also print the select result to client console through the `TableResult.print()` method. The result data in `TableResult` can be accessed only once, so for a `TableResult` instance, `collect()` method and `print()` method can't be called both.
```
Comment (twalthr) [inline docs/dev/table/sql/queries.md:146]:
```suggestion
We can also print the select result to client console through the `TableResult.print()` method. The result data in `TableResult` can be accessed only once. Thus, `collect()` and `print()` must not be called after each other.
```

File: docs/dev/table/tableApi.md [hunk 33]
```diff
@@ -2060,13 +2060,13 @@ result3 = in.order_by("a.asc").offset(10).fetch(5)
         <span class="label label-primary">Batch</span> <span class="label label-primary">Streaming</span>
       </td>
       <td>
-        <p>Similar to the INSERT INTO clause in a SQL query. Performs a insertion into a registered output table.</p>
+        <p>Similar to the INSERT INTO clause in a SQL query. Performs a insertion into a registered output table. The executeInsert method will immediately submit a flink job which execute the insert operation.</p>
```
Comment (twalthr) [inline docs/dev/table/tableApi.md:2063]:
```suggestion
        <p>Similar to the `INSERT INTO` clause in a SQL query, the method performs an insertion into a registered output table. The `executeInsert()` method will immediately submit a Flink job which execute the insert operation.</p>
```

**Comment by godfreyhe** [comment]
@twalthr Thanks for the detailed review, I have updated the pr based on comments.

**Comment by twalthr** [comment]
Thanks @godfreyhe. Merging...

**Comment by godfreyhe** [comment]
merged (see [FLINK-17599](https://issues.apache.org/jira/browse/FLINK-17599?focusedCommentId=17143097&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17143097))
